### PR TITLE
Metadata support for safetensors

### LINKED ISSUE #633: [Feature request] Add metadata to safetensors saving and loading
Just found the model weight saved in SafeTensor by mlx cannot be loaded by Transformers. An error is thrown, stating that the format is not supported. This issue seems to caused from the metadata format being saved as `mlx` (https://github.com/ml-explore/mlx/blob/main/mlx/io/safetensor.cpp#L165). However, mlx has no trouble loading HF model weights. so, I'm curious about the differences in the saved weights between them. 
Thanks in advance.

### PR DESCRIPTION
**Pull request by dc-dc-dc**
## Proposed changes

Adds metadata support for safetensors, Closes #633 

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by dc-dc-dc** [comment]
added some aliases so its a little more readable, but if not desired i can undo that commit

File: python/src/ops.cpp [hunk 1]
```diff
@@ -3225,6 +3226,7 @@ void init_ops(py::module_& m) {
         Args:
             file (file, str): File in which the array is saved.
             arrays (dict(str, array)): The dictionary of names to arrays to be saved.
+            metadata (dict(str, str)): The dictionary of metadata to be saved.
```
Comment (awni) [inline python/src/ops.cpp:3229]:
```suggestion
            metadata (dict(str, str), optional): The dictionary of metadata to be saved.
```

File: mlx/io/safetensor.cpp [hunk 2]
```diff
@@ -204,7 +208,8 @@ void save_safetensors(
 
 void save_safetensors(
     const std::string& file_,
-    std::unordered_map<std::string, array> a) {
+    std::unordered_map<std::string, array> a,
+    std::unordered_map<std::string, std::string> metadata) {
```
Comment (awni) [inline mlx/io/safetensor.cpp:212]:
```suggestion
    std::unordered_map<std::string, std::string> metadata /* = {} */) {
```

**Review by awni — COMMENTED** [review]
This looks really great, thanks for the quick addition! Mostly cosmetic suggestions, let me know when done and we can merge it!

**Comment by dc-dc-dc** [comment]
addressed the comments in the latest commit, not finding instances of `safe_tensors` what file is it in?

**Comment by awni** [comment]
> not finding instances of safe_tensors what file is it in

Sorry I meant we don't treat is as two words (e.g. we don't do `safe_tensors` we do `safetensors`

**Comment by awni** [comment]
@dc-dc-dc would you mind adding one more test in python to cover the binding?

**Review by awni — APPROVED** [review]
Looks great, thanks!!

**Pull request merged**